### DISCUSSION TITLE
What is your "favorite" design pattern

### DISCUSSION THREAD

**Discussion by lefticus** [lefticus/cpp_weekly#145] [Polls]
Yes, I know this is a ridiculous question.

However, I'm trying to prioritize the design patterns for issue #29

**Comment by jcsahnwaldt** [comment]
Composite

**Comment by trcwm** [comment]
PubSub, read-write lock.

**Comment by lefticus** [comment]
Comments from Twitter: 

> Out of GoF i recall Flyweight, itetator and Bridge. Though suppose nobody loves Bridge.
>
> From non-GoF I liked CoW, also pimpl/envelope deserbes attention


> PubSub, Visitor, read-write lock. But I’m mostly interested in patterns I am less familiar with.


> Visitor

> Missing my favourite: interpreter. Such a wildly misunderstood set of patterns all hiding in the same name.

**Comment by oschonrock** [comment]
Strategy ?

**Comment by gracicot** [comment]
I find Factory to be one of the most useful. You take a string, a int or an enum and you return an instance of a type. I use it with variant (factory can be non polymorphic and exhaustive right?) and quite useful to parse some data and output different types.

**Comment by bernd5** [comment]
Visitor pattern for ASTs or similar

**Comment by villevoutilainen** [comment]
Strategy, Template Method, and Bridge seem like unfortunate omissions, to me. The combination of the first two brings forth aspect-oriented programming. The latter, despite someone doubting it on twitter, is massively useful for ABI insulation.

**Reply by villevoutilainen** [reply]
Disclaimer: I have just used these patterns in practical industry projects, I'm not talking purely theoretically. :)

**Reply by lefticus** [reply]
Main problem is just how many github let me put in the poll. But I think I figured out how to add more, so I might rerun this

**Comment by villevoutilainen** [comment]
Oh, and for heaven's sake, the list is missing State.
